File: algorithms/3_hard/W/00212_word_search_ii/trie.cpp
typedef ptrdiff_t isize;

constexpr char INVALID_CHAR = '#';

struct Node {
    isize childs[26];
    bool is_end;
    bool is_deleted;

    Node() {
       for (auto& c : childs) { c = -1; }
       is_end = false;
       is_deleted = false;
    }
};

struct Trie {
    vector<Node> arr;
    Trie(){}

    inline static void init_freqs_board(span<size_t> freqs, const vector<vector<char>>& board) {
        for (auto& r : board) {
            for (const unsigned char c : r) {
                assert(c >= 'a' && c <= 'z');
                freqs[c - 'a'] += 1;
            }
        }
    }

   inline static void init_adj_board(span<size_t> adj, const vector<vector<char>>& board) {
        const size_t rows = board.size();
        const size_t cols = board[0].size();
        for (size_t i = 0; i < rows; i += 1) {
            for (size_t j = 0; j < cols; j += 1) {
                const size_t curr = static_cast<size_t>(board[i][j] - 'a');
                if (i > 0) {
                    const size_t up = static_cast<size_t>(board[i - 1][j] - 'a');
                    adj[up * 26 + curr] += 1;
                    if (up != curr) { adj[curr * 26 + up] += 1; }
                }

                if (j > 0) {
                    const size_t le = static_cast<size_t>(board[i][j - 1] - 'a');
                    adj[le * 26 + curr] += 1;
                    if (le != curr) { adj[curr * 26 + le] += 1; }
                }
            }
        }
    }

    static Trie from_string_slices(const vector<string>& words, const vector<vector<char>>& board) {
        size_t freqs_board[26] = { 0 };
        init_freqs_board(freqs_board, board);
        size_t adj_board [26 * 26] = { 0 };
        init_adj_board(adj_board, board);

        auto arr = vector<Node>();
        arr.push_back(Node());
        for (const string& word : words) {
            bool is_valid = true;
            size_t freqs_word[26] = { 0 };
            for (const unsigned char ch : word) {
                assert(ch >= 'a' && ch <= 'z');
                freqs_word[ch - 'a'] += 1;
                if (freqs_word[ch - 'a'] > freqs_board[ch - 'a']) {
                    is_valid = false;
                    break;
                }
            }
            if (!is_valid) { continue; }

            size_t adj_word[26 * 26] = { 0 };
            const size_t word_len = word.size();
            for (size_t i = 1; i < word_len; i += 1) {
                const size_t a = static_cast<size_t>(word[i] - 'a');
                const size_t b = static_cast<size_t>(word[i - 1] - 'a');
                adj_word[a * 26 + b] += 1;
                if (a != b) {
                    adj_word[b * 26 + a] += 1;
                }
                if (adj_word[a * 26 + b] > adj_board[a * 26 + b]) {
                    is_valid = false;
                    break;
                }
            }
            if (!is_valid) { continue; }

            size_t curr = 0;
            for (const unsigned char ch : word) {
                const size_t child_idx  = static_cast<size_t>(ch - 'a');
                if (arr[curr].childs[child_idx] == -1) {
                    arr[curr].childs[child_idx] = std::ssize(arr);
                    arr.push_back(Node());
                }
                curr = static_cast<size_t>(arr[curr].childs[child_idx]);
            }
            arr[curr].is_end = true;
        }

        auto res = Trie();
        res.arr = arr;
        return res;
    }

    inline bool mark_node_deleted_if_possible(const size_t node_idx) {
        if (arr[node_idx].is_end) { return false; }

        for (const auto& c : arr[node_idx].childs) {
            if (c != -1 && !arr[c].is_deleted) {
                return false;
            }
        }

        arr[node_idx].is_deleted = true;
        return true;
    }

    void dfs(
        const size_t trie_idx,
        vector<vector<char>>& board,
        const isize rows,
        const isize cols,
        const isize i,
        const isize j,
        string& tmp,
        vector<string>& res
    ) {
        if (arr[trie_idx].is_deleted) { return; }

        if (arr[trie_idx].is_end) {
            arr[trie_idx].is_end = false;
            res.push_back(tmp);
            if (mark_node_deleted_if_possible(trie_idx)) { return; }
        }

        if (i < 0 || i >= rows || j < 0 || j >= cols || board[i][j] == INVALID_CHAR) {
            return;
        }

        const char orig = board[i][j];
        const isize new_trie_idx_isize = arr[trie_idx].childs[static_cast<unsigned char>(orig - 'a')];
        if (new_trie_idx_isize == -1) {
            return;
        }

        const size_t new_trie_idx = static_cast<size_t>(new_trie_idx_isize);
        board[i][j] = INVALID_CHAR;
        tmp.push_back(orig);

        dfs(new_trie_idx, board, rows, cols, i + 1, j, tmp, res);
        dfs(new_trie_idx, board, rows, cols, i - 1, j, tmp, res);
        dfs(new_trie_idx, board, rows, cols, i, j + 1, tmp, res);
        dfs(new_trie_idx, board, rows, cols, i, j - 1, tmp, res);
        mark_node_deleted_if_possible(new_trie_idx);

        tmp.pop_back();
        board[i][j] = orig;
    }

    inline static vector<string> run(vector<vector<char>>& board, const vector<string>& words) {
        const isize rows = std::ssize(board);
        const isize cols = std::ssize(board[0]);
        auto trie = from_string_slices(words, board);
        auto tmp = string();
        auto res = vector<string>();

        for (isize i = 0; i < rows; i += 1) {
            for (isize j = 0; j < cols; j += 1) {
                trie.dfs(0, board, rows, cols, i, j, tmp, res);
                if (trie.mark_node_deleted_if_possible(0)) { return res; }
            }
        }

        return res;
    }
};

class Solution {
public:
    static vector<string> findWords(vector<vector<char>>& board, const vector<string>& words) {
        return Trie::run(board, words);
    }
};

File: algorithms/3_hard/W/00212_word_search_ii/trie.rs
const INVALID_CHAR: char = '#';

struct Node {
    childs: [isize; 26],
    is_end: bool,
    is_deleted: bool,
}

impl Node {
    fn new() -> Self {
        Self {
            childs: [-1; 26],
            is_end: false,
            is_deleted: false,
        }
    }
}

struct Trie {
    arr: Vec<Node>,
}

impl Trie {
    #[inline(always)]
    fn init_freqs_board(freqs: &mut [usize], board: &[Vec<char>]) {
        for r in board {
            for &c in r {
                if (c as u8) < b'a' || (c as u8) > b'z' {
                    panic!("invalid input");
                }
                freqs[((c as u8) - b'a') as usize] += 1;
            }
        }
    }

    #[inline(always)]
    fn init_adj_board(adj: &mut [usize], board: &[Vec<char>]) {
        let rows = board.len();
        let cols = board[0].len();
        for i in 0..rows {
            for j in 0..cols {
                let curr = ((board[i][j] as u8) - b'a') as usize;
                if i > 0 {
                    let up = ((board[i - 1][j] as u8) - b'a') as usize;
                    adj[up * 26 + curr] += 1;
                    if up != curr { adj[curr * 26 + up] += 1; }
                }

                if j > 0 {
                    let le = ((board[i][j - 1] as u8) - b'a') as usize;
                    adj[le * 26 + curr] += 1;
                    if le != curr { adj[curr * 26 + le] += 1; } 
                }
            }
        }
    }

    fn from_string_slices(words: &[String], board: &[Vec<char>]) -> Self {
        let mut freqs_board = [0_usize; 26];
        Self::init_freqs_board(&mut freqs_board, board);
        let mut adj_board = [0_usize; 26 * 26];
        Self::init_adj_board(&mut adj_board, board);

        let mut arr: Vec<Node> = Vec::new();
        arr.push(Node::new());
        'main_loop: for word in words {
            let mut freqs_word = [0_usize; 26];
            for &b in word.as_bytes() {
                if b < b'a' || b > b'z' {
                    panic!("invalid input");
                }
                freqs_word[(b - b'a') as usize] += 1;
                if freqs_word[(b - b'a') as usize] > freqs_board[(b - b'a') as usize] {
                    continue 'main_loop;
                }
            }

            let mut adj_word = [0_usize; 26 * 26];
            for w in word.as_bytes().windows(2) {
                let a = (w[0] - b'a') as usize;
                let b = (w[1] - b'a') as usize;
                adj_word[a * 26 + b] += 1;
                if a != b {
                    adj_word[b * 26 + a] += 1;
                }
                if adj_word[a * 26 + b] > adj_board[a * 26 + b] {
                    continue 'main_loop;
                }
            }
            
            let mut curr = 0_usize;
            for &b in word.as_bytes() {
                let child_idx = (b - b'a') as usize;
                if arr[curr].childs[child_idx] == -1 {
                    arr[curr].childs[child_idx] = arr.len() as isize;
                    arr.push(Node::new());
                }
                curr = arr[curr].childs[child_idx] as usize;
            }
            arr[curr].is_end = true;
        }

        Self {
            arr,
        }
    }

    #[inline(always)]
    fn mark_node_deleted_if_possible(&mut self, node_idx: usize) -> bool {
        if self.arr[node_idx].is_end { return false; }

        for &c in &self.arr[node_idx].childs {
            if c != -1 && !self.arr[c as usize].is_deleted {
                return false;
            }
        }

        self.arr[node_idx].is_deleted = true;
        true
    }

    fn dfs(
        &mut self,
        trie_idx: usize,
        board: &mut [Vec<char>],
        cols: isize,
        i: isize,
        j: isize,
        tmp: &mut Vec<u8>,
        res: &mut Vec<String>,
    ) {
        if self.arr[trie_idx].is_deleted { return; }

        if self.arr[trie_idx].is_end {
            self.arr[trie_idx].is_end = false;
            res.push(String::from_utf8_lossy(tmp).into_owned());
            if self.mark_node_deleted_if_possible(trie_idx) { return; }
        }

        let rows = board.len() as isize;
        if i < 0 || i >= rows || j < 0 || j >= cols || board[i as usize][j as usize] == INVALID_CHAR {
            return;
        }

        let orig = board[i as usize][j as usize];
        let new_trie_idx_isize = self.arr[trie_idx].childs[((orig as u8) - b'a') as usize];
        if new_trie_idx_isize == -1 {
            return;
        }

        let new_trie_idx = new_trie_idx_isize as usize;
        board[i as usize][j as usize] = INVALID_CHAR;
        tmp.push(orig as u8);

        self.dfs(new_trie_idx, board, cols, i + 1, j, tmp, res);
        self.dfs(new_trie_idx, board, cols, i - 1, j, tmp, res);
        self.dfs(new_trie_idx, board, cols, i, j + 1, tmp, res);
        self.dfs(new_trie_idx, board, cols, i, j - 1, tmp, res);
        let _ = self.mark_node_deleted_if_possible(new_trie_idx);

        tmp.pop();
        board[i as usize][j as usize] = orig;
    }

    #[inline(always)]
    pub fn run(board: &mut [Vec<char>], words: &[String]) -> Vec<String> {
        let rows = board.len() as isize;
        let cols = board[0].len() as isize;
        let mut trie = Self::from_string_slices(words, board);
        let mut tmp: Vec<u8> = Vec::new();
        let mut res: Vec<String> = Vec::new();

        'main_loop: for i in 0..rows {
            for j in 0..cols {
                trie.dfs(0, board, cols, i, j, &mut tmp, &mut res);
                if trie.mark_node_deleted_if_possible(0) { break 'main_loop; }
            }
        }

        res
    }
}

impl Solution {
    pub fn find_words(mut board: Vec<Vec<char>>, words: Vec<String>) -> Vec<String> {
        Trie::run(&mut board, &words)
    }
}
